### PR TITLE
Add the basic support of grpclb

### PR DESCRIPTION
**Pull request by iamqizhao**
https://github.com/grpc/grpc/blob/master/doc/load-balancing.md

### CODE REVIEW

**Comment by iamqizhao** [comment]
\cc @xiang90 I need to tune the Balancer API a bit to accommodate this change for security reasons.

File: grpclb/grpclb.go [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+	waitCh chan struct{}
+	done   bool
+}
+
+func (b *balancer) watchAddrUpdates(w naming.Watcher, ch chan remoteBalancerInfo) error {
+	updates, err := w.Next()
+	if err != nil {
+		return err
+	}
+	var bAddr remoteBalancerInfo
```
Comment (iamqizhao) [inline grpclb/grpclb.go]: done

File: grpclb/grpclb.go [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+	for _, update := range updates {
+		addr := grpc.Address{
+			Addr:     update.Addr,
+			Metadata: update.Metadata,
+		}
+		switch update.Op {
+		case naming.Add:
+			var exist bool
+			for _, v := range b.rbs {
+				// TODO: Is the same addr with different different server name a different balancer?
```
Comment (iamqizhao) [inline grpclb/grpclb.go]: done.

File: grpclb/grpclb.go [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+func (b *balancer) Notify() <-chan []grpc.Address {
+	return b.addrCh
+}
+
+func (b *balancer) Close() error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.done = true
+	if b.waitCh != nil {
+		close(b.waitCh)
```
Comment (iamqizhao) [inline grpclb/grpclb.go]: The code always checks b.done before checking b.waitCh now.

File: grpclb/grpclb_test.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
+/*
+ *
+ * Copyright 2014, Google Inc.
```
Comment (iamqizhao) [inline grpclb/grpclb_test.go]: done

File: grpclb/grpclb_test.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+	// the channel to receives name resolution updates
+	update chan *naming.Update
+	// the side channel to get to know how many updates in a batch
+	side chan int
+	// the channel to notifiy update injector that the update reading is done
+	readDone chan int
+}
+
+func (w *testWatcher) Next() (updates []*naming.Update, err error) {
+	n := <-w.side
```
Comment (iamqizhao) [inline grpclb/grpclb_test.go]: done

File: grpclb/grpclb_test.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+	if n == 0 {
+		return nil, fmt.Errorf("w.side is closed")
+	}
+	for i := 0; i < n; i++ {
+		u := <-w.update
```
Comment (iamqizhao) [inline grpclb/grpclb_test.go]: done

File: grpclb/grpclb_test.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+		}
+	}
+	w.readDone <- 0
+	return
+}
+
+func (w *testWatcher) Close() {
+}
+
+// Inject naming resolution updates to the testWatcher.
```
Comment (iamqizhao) [inline grpclb/grpclb_test.go]: this is just a comment of an unexported function :)

File: grpclb/grpclb_test.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+	return
+}
+
+func stopBackends(servers []*grpc.Server) {
+	for _, s := range servers {
+		s.Stop()
+	}
+}
+
+func TestGrpcLB(t *testing.T) {
```
Comment (iamqizhao) [inline grpclb/grpclb_test.go]: done

**Comment by akaspin** [comment]
Thanks pal. You are steal two hours of my life.

**Pull request merged**